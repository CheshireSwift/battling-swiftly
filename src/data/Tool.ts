export enum Tool {
  Measure,
  Move,
}

export default Tool
