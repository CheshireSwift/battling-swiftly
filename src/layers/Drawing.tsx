import * as React from 'react'
import * as _ from 'lodash'

import { layerStyle } from '../helpers/styleSnippets'
import { useEventValue } from '../helpers/hooks'
import Vector from '../helpers/Vector'
import PopupMenu from '../ui/PopupMenu'
import Tool from '../data/Tool'
import CharacterMarker, { Character } from '../graphics/CharacterMarker'

type DrawingProps = {
  dimensions: { height: number; width: number }
  dpi: number
  characters: Character[]
  controlledCharacterKey?: string
  update: (key: string, data: Partial<Character>) => void
}

const cappedDifference = (start: Vector, end: Vector, cap: number) =>
  start.difference(end).cap(cap)

export const Drawing = ({
  dimensions,
  dpi,
  characters,
  controlledCharacterKey,
  update,
}: DrawingProps) => {
  const movementCap = 6 * dpi
  const [moveStart, setMoveStart] = React.useState<Vector | null>(null)
  const [moveEnd, setMoveEnd] = React.useState<Vector | null>(null)
  const [menuPosition, setMenuPosition] = React.useState<Vector | null>(null)
  const [selectedTool, setSelectedTool] = React.useState(Tool.Measure)
  const mousePos = useEventValue<MouseEvent, Vector>(
    'mousemove',
    Vector.fromEvent,
  )
  const controlledCharacterPosition = () => {
    const controlledCharacter = _.find(characters, {
      key: controlledCharacterKey,
    })
    if (!controlledCharacter) {
      return
    }

    return Vector.fromXY(controlledCharacter.position)
  }

  const startToUse = {
    [Tool.Measure]: moveStart,
    [Tool.Move]: controlledCharacterPosition(),
  }[selectedTool]

  const endToUse = moveEnd || mousePos

  const measureHandler = (e: React.MouseEvent): void => {
    const point = Vector.fromEvent(e)
    if (moveStart && !moveEnd) {
      setMoveEnd(point)
    } else {
      setMoveStart(point)
      setMoveEnd(null)
    }
  }

  const moveHandler = (e: React.MouseEvent): void => {
    const position = controlledCharacterPosition()
    controlledCharacterKey &&
      position &&
      update(controlledCharacterKey, {
        position: cappedDifference(position, Vector.fromEvent(e), movementCap)
          .add(position)
          .toXY(),
      })

    setSelectedTool(Tool.Measure)
  }

  const handlerForTool = (tool: Tool) => {
    switch (tool) {
      case Tool.Measure:
        return measureHandler
      case Tool.Move:
        return moveHandler
      default:
        console.warn('No handler for tool', Tool[tool])
        return () => {}
    }
  }

  const toolClick = (e: React.MouseEvent) => {
    if (menuPosition) {
      setMenuPosition(null)
      return
    }

    if (e.button !== 0) {
      return
    }

    handlerForTool(selectedTool)(e)
  }

  const menuClick = (e: React.MouseEvent) => {
    e.preventDefault()
    setMoveStart(null)
    setMoveEnd(null)
    setMenuPosition(Vector.fromEvent(e))
  }

  const highlightedChildren = characters.map(character => (
    <CharacterMarker
      key={character.key}
      character={character}
      highlight={
        selectedTool == Tool.Move && character.key === controlledCharacterKey
      }
      dpi={dpi}
    />
  ))

  const plottingLine = () => {
    if (!startToUse || !endToUse) {
      return null
    }

    if (selectedTool === Tool.Measure) {
      return <PlottingLine start={startToUse} end={endToUse} dpi={dpi} />
    }

    if (selectedTool === Tool.Move) {
      const lineVector = startToUse.difference(endToUse)
      const scaledLineVector = lineVector.cap(movementCap)

      return (
        <PlottingLine
          start={startToUse}
          end={startToUse.add(scaledLineVector)}
          dpi={dpi}
        />
      )
    }

    return null
  }

  return (
    <div onClick={toolClick} onContextMenu={menuClick}>
      <svg style={{ ...layerStyle, ...dimensions }}>
        {highlightedChildren}
        {plottingLine()}
      </svg>
      <PopupMenu
        position={menuPosition ? menuPosition.toStyle() : null}
        menuItems={[
          [Tool.Measure, 'Measure'],
          controlledCharacterKey && [Tool.Move, 'Move'],
        ]}
        selectedItem={selectedTool}
        onSelectItem={(selection: Tool) => {
          setSelectedTool(selection)
          setMenuPosition(null)
        }}
      />
    </div>
  )
}

type PlottingLineProps = {
  start: Vector
  end: Vector
  dpi: number
}
const PlottingLine = ({ start, end, dpi }: PlottingLineProps) => {
  const lineVector = start.difference(end)
  return (
    <>
      <line
        {...start.suffix('1')}
        {...end.suffix('2')}
        stroke="lime"
        strokeWidth="1.5"
      />
      <circle
        {...end.prefix('c')}
        r="4"
        stroke="lime"
        strokeWidth="1.5"
        fill="none"
      />
      <text
        {...end}
        fill="lime"
        fontSize="16"
        textAnchor="middle"
        dominantBaseline="middle"
        {...lineVector.normalize(30, 15).prefix('d')}
        style={{
          fontFamily: 'monospace',
          fontWeight: 'bold',
          cursor: 'pointer',
        }}
      >
        {(lineVector.length() / dpi).toFixed(1)}"
      </text>
    </>
  )
}
export default Drawing
