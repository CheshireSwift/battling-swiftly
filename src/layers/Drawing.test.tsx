import * as React from 'react'
import { mount, render, ReactWrapper } from 'enzyme'

import Drawing from './Drawing'
import Tool from '../data/Tool'

function selectTool(drawingLayer: ReactWrapper, tool: Tool) {
  drawingLayer.simulate('contextmenu')
  drawingLayer.find('#popup-menu-item-' + tool).simulate('click')
}

let entropy = 0
const character = (name: string) => ({
  name,
  key: '' + entropy++,
  position: { x: entropy++, y: entropy++ },
})

describe('the drawing layer', () => {
  it('matches the snapshot', () => {
    expect(
      render(
        <Drawing
          dimensions={{ width: 200, height: 100 }}
          dpi={10}
          characters={[
            { name: 'Bob', key: '123xyz', position: { x: 1, y: 2 } },
          ]}
          update={jest.fn}
        />,
      ),
    ).toMatchSnapshot()
  })

  it('allows rangefinding', () => {
    const drawingLayer = mount(
      <Drawing
        dimensions={{ width: 200, height: 100 }}
        dpi={10}
        characters={[]}
        update={jest.fn}
      />,
    )

    drawingLayer.simulate('click', { button: 0, pageX: 10, pageY: 20 })
    drawingLayer.simulate('click', { button: 0, pageX: 40, pageY: 60 })

    expect(drawingLayer.find('line').prop('x1')).toBe(10)
    expect(drawingLayer.find('line').prop('y1')).toBe(20)

    expect(drawingLayer.find('line').prop('x2')).toBe(40)
    expect(drawingLayer.find('line').prop('y2')).toBe(60)

    expect(drawingLayer.text()).toContain('5.0"')
    expect(drawingLayer.text()).toContain('4,6')
  })

  it('shows the menu on right click', () => {
    const drawingLayer = mount(
      <Drawing
        dimensions={{ width: 20, height: 10 }}
        dpi={1}
        characters={[]}
        update={jest.fn()}
      />,
    )

    drawingLayer.simulate('contextmenu', { pageX: 10, pagetY: 20 })

    expect(drawingLayer.text()).toContain('Measure')
  })

  it('renders its characters', () => {
    const drawingLayer = render(
      <Drawing
        dimensions={{ width: 200, height: 100 }}
        dpi={1}
        characters={[character('Waffle'), character('Pickle')]}
        update={jest.fn()}
      />,
    )

    expect(drawingLayer.text()).toContain('Waffle')
    expect(drawingLayer.text()).toContain('Pickle')
  })

  it('removes the movement tool when no character is being controlled', () => {
    const drawingLayer = mount(
      <Drawing
        dimensions={{ width: 200, height: 100 }}
        dpi={1}
        characters={[]}
        update={jest.fn()}
      />,
    )
    drawingLayer.simulate('contextmenu')
    expect(drawingLayer.text()).not.toContain('Move')
  })

  it('highlights the character that matches the controlled key when the movement tool is selected', () => {
    const hero = character('Zero')
    const other = character('X')
    const drawingLayer = mount(
      <Drawing
        dimensions={{ width: 200, height: 100 }}
        dpi={1}
        controlledCharacterKey={hero.key}
        characters={[hero, other]}
        update={jest.fn()}
      />,
    )

    selectTool(drawingLayer, Tool.Move)

    expect(
      drawingLayer.findWhere(e => e.key() === hero.key).prop('highlight'),
    ).toBeTruthy()

    expect(
      drawingLayer.findWhere(e => e.key() === other.key).prop('highlight'),
    ).toBeFalsy()
  })

  it('fires an update when the controlled character is moved', () => {
    const hero = { key: '1234', name: 'Zero', position: { x: 3, y: 4 } }
    const update = jest.fn()
    const drawingLayer = mount(
      <Drawing
        dimensions={{ width: 20, height: 20 }}
        dpi={1}
        controlledCharacterKey={hero.key}
        characters={[hero]}
        update={update}
      />,
    )

    selectTool(drawingLayer, Tool.Move)

    drawingLayer.simulate('click', { button: 0, pageX: 5, pageY: 6 })

    expect(update).toHaveBeenCalledWith(hero.key, { position: { x: 5, y: 6 } })
  })

  it('caps the update with the movement limit (6")', () => {
    const hero = { key: '1234', name: 'Zero', position: { x: 3, y: 4 } }
    const update = jest.fn()
    const drawingLayer = mount(
      <Drawing
        dimensions={{ width: 20, height: 20 }}
        dpi={1}
        controlledCharacterKey={hero.key}
        characters={[hero]}
        update={update}
      />,
    )

    selectTool(drawingLayer, Tool.Move)

    drawingLayer.simulate('click', { button: 0, pageX: 3, pageY: 20 })

    expect(update).toHaveBeenCalledWith(hero.key, {
      position: { x: 3, y: 4 + 6 },
    })
  })
})
