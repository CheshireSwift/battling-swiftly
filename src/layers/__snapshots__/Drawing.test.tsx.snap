// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the drawing layer matches the snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "style": "position:absolute;top:0;bottom:0;width:200px;height:100px",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "cx": "1",
              "cy": "2",
              "fill": "lime",
              "fill-opacity": "0.1",
              "r": "4",
              "stroke": "lime",
              "stroke-width": "1.5",
            },
            "children": Array [],
            "name": "circle",
            "namespace": "http://www.w3.org/2000/svg",
            "next": Object {
              "attribs": Object {
                "d": "
M 3,0
L 16,-11
L 34,-11
    ",
                "fill": "none",
                "stroke": "lime",
                "stroke-width": "1.5",
              },
              "children": Array [],
              "name": "path",
              "namespace": "http://www.w3.org/2000/svg",
              "next": Object {
                "attribs": Object {
                  "fill": "lime",
                  "style": "font-family:monospace;font-weight:bold;font-size:14px",
                  "x": "16",
                  "y": "-13",
                },
                "children": Array [
                  Object {
                    "data": "Bob",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "text",
                "namespace": "http://www.w3.org/2000/svg",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "fill": undefined,
                  "style": undefined,
                  "x": undefined,
                  "y": undefined,
                },
                "x-attribsPrefix": Object {
                  "fill": undefined,
                  "style": undefined,
                  "x": undefined,
                  "y": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "d": undefined,
                "fill": undefined,
                "stroke": undefined,
                "stroke-width": undefined,
              },
              "x-attribsPrefix": Object {
                "d": undefined,
                "fill": undefined,
                "stroke": undefined,
                "stroke-width": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "cx": undefined,
              "cy": undefined,
              "fill": undefined,
              "fill-opacity": undefined,
              "r": undefined,
              "stroke": undefined,
              "stroke-width": undefined,
            },
            "x-attribsPrefix": Object {
              "cx": undefined,
              "cy": undefined,
              "fill": undefined,
              "fill-opacity": undefined,
              "r": undefined,
              "stroke": undefined,
              "stroke-width": undefined,
            },
          },
          Object {
            "attribs": Object {
              "d": "
M 3,0
L 16,-11
L 34,-11
    ",
              "fill": "none",
              "stroke": "lime",
              "stroke-width": "1.5",
            },
            "children": Array [],
            "name": "path",
            "namespace": "http://www.w3.org/2000/svg",
            "next": Object {
              "attribs": Object {
                "fill": "lime",
                "style": "font-family:monospace;font-weight:bold;font-size:14px",
                "x": "16",
                "y": "-13",
              },
              "children": Array [
                Object {
                  "data": "Bob",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "text",
              "namespace": "http://www.w3.org/2000/svg",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "fill": undefined,
                "style": undefined,
                "x": undefined,
                "y": undefined,
              },
              "x-attribsPrefix": Object {
                "fill": undefined,
                "style": undefined,
                "x": undefined,
                "y": undefined,
              },
            },
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "cx": "1",
                "cy": "2",
                "fill": "lime",
                "fill-opacity": "0.1",
                "r": "4",
                "stroke": "lime",
                "stroke-width": "1.5",
              },
              "children": Array [],
              "name": "circle",
              "namespace": "http://www.w3.org/2000/svg",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "cx": undefined,
                "cy": undefined,
                "fill": undefined,
                "fill-opacity": undefined,
                "r": undefined,
                "stroke": undefined,
                "stroke-width": undefined,
              },
              "x-attribsPrefix": Object {
                "cx": undefined,
                "cy": undefined,
                "fill": undefined,
                "fill-opacity": undefined,
                "r": undefined,
                "stroke": undefined,
                "stroke-width": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "d": undefined,
              "fill": undefined,
              "stroke": undefined,
              "stroke-width": undefined,
            },
            "x-attribsPrefix": Object {
              "d": undefined,
              "fill": undefined,
              "stroke": undefined,
              "stroke-width": undefined,
            },
          },
          Object {
            "attribs": Object {
              "fill": "lime",
              "style": "font-family:monospace;font-weight:bold;font-size:14px",
              "x": "16",
              "y": "-13",
            },
            "children": Array [
              Object {
                "data": "Bob",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "text",
            "namespace": "http://www.w3.org/2000/svg",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "d": "
M 3,0
L 16,-11
L 34,-11
    ",
                "fill": "none",
                "stroke": "lime",
                "stroke-width": "1.5",
              },
              "children": Array [],
              "name": "path",
              "namespace": "http://www.w3.org/2000/svg",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "cx": "1",
                  "cy": "2",
                  "fill": "lime",
                  "fill-opacity": "0.1",
                  "r": "4",
                  "stroke": "lime",
                  "stroke-width": "1.5",
                },
                "children": Array [],
                "name": "circle",
                "namespace": "http://www.w3.org/2000/svg",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "cx": undefined,
                  "cy": undefined,
                  "fill": undefined,
                  "fill-opacity": undefined,
                  "r": undefined,
                  "stroke": undefined,
                  "stroke-width": undefined,
                },
                "x-attribsPrefix": Object {
                  "cx": undefined,
                  "cy": undefined,
                  "fill": undefined,
                  "fill-opacity": undefined,
                  "r": undefined,
                  "stroke": undefined,
                  "stroke-width": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "d": undefined,
                "fill": undefined,
                "stroke": undefined,
                "stroke-width": undefined,
              },
              "x-attribsPrefix": Object {
                "d": undefined,
                "fill": undefined,
                "stroke": undefined,
                "stroke-width": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "fill": undefined,
              "style": undefined,
              "x": undefined,
              "y": undefined,
            },
            "x-attribsPrefix": Object {
              "fill": undefined,
              "style": undefined,
              "x": undefined,
              "y": undefined,
            },
          },
        ],
        "name": "svg",
        "namespace": "http://www.w3.org/2000/svg",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
